### ISSUE TITLE
Race condition can lead Pod to "steal" other Pod's PVC after CSI migration

### ISSUE THREAD

**Issue by ialidzhikov** [kubernetes/kubernetes#102856]
/sig storage
/king bug

#### What happened:

While upgrading a K8s cluster from v1.17.x to 1.18.18 (with CSI migration enabled), we ended up with a Pod that managed to "steal" and mount another Pod's PVC.

We have a Pod `pod-1` that according to its spec needs to mount `vol-1` and Pod `pod-2` that needs to mount `vol-2`. We also have a Node `worker` (backed by instance id `i-1234`, machine type `x1.16xlarge`). In original, this issue observed in a cluster with ~2000 PVs and ~120 Nodes.

After an investigation, we built the following chronological order for the corresponding 2 Pods and 2 volumes:

1. (12:29:27) `pod-1` is scheduled on Node `worker`

kube-scheduler:

```
2021-06-05 12:29:27 {"log":"pod default/pod-1 is bound successfully on node \"worker\", 132 nodes evaluated, 3 nodes were found feasible.","pid":"1","severity":"INFO","source":"scheduler.go:742"}
```

2. (12:29:59) `pod-2` is scheduled on Node `worker`

kube-scheduler:

```
2021-06-05 12:29:59 {"log":"pod default/pod-2 is bound successfully on node \"worker\", 132 nodes evaluated, 3 nodes were found feasible.","pid":"1","severity":"INFO","source":"scheduler.go:742"}
```

3. (12:30:53) `vol-1` is attached to Node `worker` (backing instance id escaped as `i-1234`) (device `/dev/xvdba`) by the aws-ebs-csi-driver

aws-ebs-csi-driver:

```
2021-06-05 12:30:53	{"log":"ControllerPublishVolume: volume vol-1 attached to node i-1234 through device /dev/xvdba","pid":"1","severity":"INFO","source":"controller.go:322"}
```

4. (12:32:02 - 12:34:01) `vol-1` successfully **(and wrongly)** detached by kube-controller-manager from `i-1234`

```
2021-06-05 12:31:29 {"log":"Marking volume attachment as uncertain as volume:\"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\" (\"worker\") is not attached (Detached)","pid":"1","severity":"INFO","source":"attach_detach_controller.go:757"}
2021-06-05 12:32:02 {"log":"attacherDetacher.DetachVolume started for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\") on node \"worker\" ","pid":"1","severity":"INFO","source":"reconciler.go:219"}
2021-06-05 12:34:01 {"log":"DetachVolume.Detach succeeded for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\") on node \"worker\" ","pid":"1","severity":"INFO","source":"operation_generator.go:472"}
```

A normal detach (i.e not a force-detach) was successful because the volume `vol-1` was still not mounted on the Node.

kubelet:

```
2021-06-05T12:31:35.817863+00:00 localhost kubelet[6458]: E0605 12:31:35.817799    6458 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/ebs.csi.aws.com^vol-1 podName: nodeName:}" failed. No retries permitted until 2021-06-05 12:33:37.817769943 +0000 UTC m=+3741.823630839 (durationBeforeRetry 2m2s). Error: "Volume not attached according to node status for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-1\") pod \"pod-1\" (UID: \"3efee008-b66e-44d3-b8bc-bcbbfc44229b\") "

```

5. (12:34:42) `pod-1` still fails to start because kubelet cannot find device /dev/xvdba

kubelet:

```
2021-06-05T12:34:42.423673+00:00 localhost kubelet[6458]: E0605 12:34:42.423297    6458 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/ebs.csi.aws.com^vol-1 podName: nodeName:}" failed. No retries permitted until 2021-06-05 12:35:46.4232653 +0000 UTC m=+3870.429126192 (durationBeforeRetry 1m4s). Error: "MountVolume.MountDevice failed for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/csi/ebs.csi.aws.com^vol-1\") pod \"pod-1\" (UID: \"3efee008-b66e-44d3-b8bc-bcbbfc44229b\") : rpc error: code = Internal desc = Failed to find device path /dev/xvdba. nvme path \"/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_1\" not found"
```

6. (12:35:40) `vol-2` is attached to `i-1234` (device `/dev/xvdba`) by the aws-ebs-csi-driver

aws-ebs-csi-driver:

```
2021-06-05 12:35:40	{"log":"ControllerPublishVolume: volume vol-2 attached to node i-1234 through device /dev/xvdba","pid":"1","severity":"INFO","source":"controller.go:322"}
```

7. (12:35:47) `pod-1` successfully mounts device path `/dev/xvdba` and the `pod-1` is started for the first time

```
2021-06-05T12:35:46.428374+00:00 localhost kubelet[6458]: I0605 12:35:46.428234    6458 reconciler.go:269] operationExecutor.MountVolume started for volume "pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-1") pod "pod-1" (UID: "3efee008-b66e-44d3-b8bc-bcbbfc44229b")
2021-06-05T12:35:46.428767+00:00 localhost kubelet[6458]: I0605 12:35:46.428685    6458 operation_generator.go:558] MountVolume.WaitForAttach entering for volume "pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-1") pod "pod-1" (UID: "3efee008-b66e-44d3-b8bc-bcbbfc44229b") DevicePath ""
2021-06-05T12:35:46.431979+00:00 localhost kubelet[6458]: I0605 12:35:46.431916    6458 operation_generator.go:567] MountVolume.WaitForAttach succeeded for volume "pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-1") pod "pod-1" (UID: "3efee008-b66e-44d3-b8bc-bcbbfc44229b") DevicePath "csi-9284d6bb1e0c9c43d1b133be67f250969088e2c236170773fd176bd5d0836d18"
2021-06-05T12:35:47.223895+00:00 localhost kubelet[6458]: I0605 12:35:47.223803    6458 operation_generator.go:596] MountVolume.MountDevice succeeded for volume "pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-1") pod "pod-1" (UID: "3efee008-b66e-44d3-b8bc-bcbbfc44229b") device mount path "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59/globalmount"
2021-06-05T12:35:47.420736+00:00 localhost kubelet[6458]: I0605 12:35:47.420671    6458 operation_generator.go:657] MountVolume.SetUp succeeded for volume "pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-1") pod "pod-1" (UID: "3efee008-b66e-44d3-b8bc-bcbbfc44229b")

2021-06-05T12:35:47.595521+00:00 localhost kubelet[6458]: I0605 12:35:47.595442    6458 kuberuntime_manager.go:428] No sandbox for pod "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)" can be found. Need to start a new one
2021-06-05T12:35:47.984148+00:00 localhost kubelet[6458]: I0605 12:35:47.984069    6458 kubelet.go:1944] SyncLoop (UPDATE, "api"): "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)"
2021-06-05T12:35:48.225816+00:00 localhost kubelet[6458]: I0605 12:35:48.225700    6458 kubelet.go:1982] SyncLoop (PLEG): "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)", event: &pleg.PodLifecycleEvent{ID:"3efee008-b66e-44d3-b8bc-bcbbfc44229b", Type:"ContainerStarted", Data:"39ffaa72e964b9148a9167adf42dd025fa8d1e977a875251600c0b594f5b68c2"}
2021-06-05T12:35:49.249357+00:00 localhost kubelet[6458]: I0605 12:35:49.249206    6458 kubelet.go:1982] SyncLoop (PLEG): "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)", event: &pleg.PodLifecycleEvent{ID:"3efee008-b66e-44d3-b8bc-bcbbfc44229b", Type:"ContainerDied", Data:"e0b7b67651909da0b59facb38680a351e96dc3a886a4b5e869f66b8b6079558e"}
2021-06-05T12:35:50.319959+00:00 localhost kubelet[6458]: I0605 12:35:50.319827    6458 kubelet.go:1982] SyncLoop (PLEG): "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)", event: &pleg.PodLifecycleEvent{ID:"3efee008-b66e-44d3-b8bc-bcbbfc44229b", Type:"ContainerStarted", Data:"30c39586a6cb49309c140c2fc84169746c816cd8233a43d3cadea650b6fefedc"}
2021-06-05T12:35:50.320368+00:00 localhost kubelet[6458]: I0605 12:35:50.319874    6458 kubelet.go:1982] SyncLoop (PLEG): "pod-1_default(3efee008-b66e-44d3-b8bc-bcbbfc44229b)", event: &pleg.PodLifecycleEvent{ID:"3efee008-b66e-44d3-b8bc-bcbbfc44229b", Type:"ContainerStarted", Data:"89d2a89df7e6dc9a3308b3cafd7269792f1c6adabe53921e0b117e7174587dda"}
```

8. (12:38:13) `pod-2` successfully mounts device path `/dev/xvdba` and `pod-2` is started for the first time

```
2021-06-05T12:38:13.684123+00:00 localhost kubelet[6458]: I0605 12:38:13.684069    6458 reconciler.go:224] operationExecutor.VerifyControllerAttachedVolume started for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e")
2021-06-05T12:38:13.687539+00:00 localhost kubelet[6458]: I0605 12:38:13.687481    6458 operation_generator.go:1332] Controller attach succeeded for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e") device path: ""
2021-06-05T12:38:13.785066+00:00 localhost kubelet[6458]: I0605 12:38:13.784923    6458 reconciler.go:269] operationExecutor.MountVolume started for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e")
2021-06-05T12:38:13.785348+00:00 localhost kubelet[6458]: I0605 12:38:13.785313    6458 operation_generator.go:558] MountVolume.WaitForAttach entering for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e") DevicePath ""
2021-06-05T12:38:13.788317+00:00 localhost kubelet[6458]: I0605 12:38:13.788174    6458 operation_generator.go:567] MountVolume.WaitForAttach succeeded for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e") DevicePath "csi-cba5926c136e6e4ebdda89f0b9bcf0081bbc6658c272c3e938396e242d7fd7c3"
2021-06-05T12:38:14.120189+00:00 localhost kubelet[6458]: I0605 12:38:14.120108    6458 operation_generator.go:596] MountVolume.MountDevice succeeded for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e") device mount path "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-9229d62a-8762-446d-a3c1-0888cd8455b6/globalmount"
2021-06-05T12:38:14.423823+00:00 localhost kubelet[6458]: I0605 12:38:14.423752    6458 operation_generator.go:657] MountVolume.SetUp succeeded for volume "pvc-9229d62a-8762-446d-a3c1-0888cd8455b6" (UniqueName: "kubernetes.io/csi/ebs.csi.aws.com^vol-2") pod "pod-2" (UID: "8b387fad-ca6f-4308-931f-24dce366437e")
2021-06-05T12:38:14.695434+00:00 localhost kubelet[6458]: I0605 12:38:14.695381    6458 kuberuntime_manager.go:428] No sandbox for pod "pod-2_default(8b387fad-ca6f-4308-931f-24dce366437e)" can be found. Need to start a new one
```

---

tl;dr: According the the specs, `pod-1` should mount `vol-1` and `pod-2` should mount `vol-2`. From the steps above we see that both `pod-1` and `pod-2` manage to mount device `/dev/xvdba` that is backed by `vol-2`. `pod-1` is able to access the data from `vol-2`.

---

We summarized our investigation in 2 issues:
1. The first issue is at `step 4.`. It is actually the trigger for the second issue described below. So the log entry `Marking volume attachment as uncertain as volume:` pretty much describes this issue as the bug fixed by https://github.com/kubernetes/kubernetes/pull/101737. In short, because of a bug, kube-controller-manager can mark a VolumeAttachment as uncertain and this can lead to unexpected and wrong detachment of migrated PV. This issue is caused by cherry-pick of https://github.com/kubernetes/kubernetes/pull/96617 to K8s v1.18. The affected 1.18 versions by this bug are v1.18.16-v1.18.19.
2. The second issue is at `step 7.` and `step 8.`. `pod-1` was scheduled, its volume `vol-1` was attached to the Node, then detached wrongly by kube-controller-manager. Pod `pod-1` failed multiple times to mount the device path, but when the same device path was attached to another volume (`vol-2`), it managed to mount it as well. IMO this is a race condition in K8s and this issue can arise with manual detach from the cloud provider (that's why it is not strictly related to `issue 1.`). We also saw that this issue can arise when only certain AWS machine types are used because we saw that the block device naming on for example `m5.xlarge` and `x1.16xlarge` is different. 

#### What you expected to happen:

Pod to not "steal" (mount) another Pod's PVC.

#### How to reproduce it (as minimally and precisely as possible):

It is not possible to provide steps to clear reproduce as 2 race condition issues are involved.

To be able to simulate the wrong detach by kube-controller-manager and the proper timing with the attachments and mounts, we ended up with the following "artificial" steps to reproduce:

1. Create a single node (AWS machine type - `x1.16xlarge`) K8s cluster (v1.18.18, CSI enabled) 

2. Create a StatefulSet with 2 replicas

<details>

<summary>Expand for the StatetulSet spec!</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  selector:
    matchLabels:
      app: nginx
  serviceName: "nginx"
  replicas: 2
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "gp2"
      resources:
        requests:
          storage: 1Gi
```

</details>

3. Exec to the Pods and create sample files on the volumes to have an unique identifier on the volumes

```
$ k exec -ti web-0 -- sh
# cd usr/share/nginx/html
# touch web-0.txt

$ k exec -ti web-1 -- sh
# cd usr/share/nginx/html
# touch web-1.txt
```

4. Save the Pods specs and VolumeAttachment specs into `web-0-pod.yaml`, `web-1-pod.yaml`, `web-0-va.yaml`, `web-1-va.yaml`

5. Delete the StatefulSet `web`. This will delete all Pods and VolumeAttachments but will leave the PVCs.

6. Scale kube-controller-manager down. 

This step is required to prevent kube-controller-manager to maintain the volume attachments in the Node status (respectively required to have a Pod (web-0) that fails to mount a volume).

6. Create Pod `web-0`

```
$ k apply -f web-0-pod.yaml
```

As kube-controller-manager is already scaled down, the Node status does not contain information about the volume attachments. And `web-0` fails to start with:

```
Events:
  Type     Reason       Age    From     Message
  ----     ------       ----   ----     -------
  Warning  FailedMount  3m14s  kubelet  Unable to attach or mount volumes: unmounted volumes=[www], unattached volumes=[www default-token-prbtr]: timed out waiting for the condition
  Warning  FailedMount  57s    kubelet  Unable to attach or mount volumes: unmounted volumes=[www], unattached volumes=[default-token-prbtr www]: timed out waiting for the condition
```

7. Create VolumeAttachment for `web-0`

```
$ k apply -f web-0-va.yaml
```

The external-attacher will then attach the underlying volume to the EC2 instance.

8. Go to the AWS console and detach the just attached volume in `step 7.`

This step is the respective step that "simulates" how kube-controller-manager wrongly detached a migrated PV.

9. Create Pod web-1

```
$ k apply -f web-1-pod.yaml
```

10. Scale kube-controller-manager up

Then kube-controller-manager will create a new VolumeAttachment for Pod `web-1`. Both VolumeAttachments will have devicePath `/dev/xvdba` (as the previous volume is already detached and `/dev/xvdba` is already free).

11. Ensure that VolumeAttachment for `web-1` is created (by kube-controller-manager) and it is attached to `/dev/xvdba`

12. Ensure that Pods `web-0` and `web-1` are successfully started and both mount the volume associated with `web-1`

```
$ k get po
NAME    READY   STATUS    RESTARTS   AGE
web-0   1/1     Running   0          10m
web-1   1/1     Running   0          8m
```

```
$ k exec -ti web-0 -- sh
# cd usr/share/nginx/html
# ls -la
total 24
drwxr-xr-x 3 root root  4096 Jun 14 09:50 .
drwxr-xr-x 4 root root  4096 Jul  1  2016 ..
drwx------ 2 root root 16384 Jun 14 09:47 lost+found
-rw-r--r-- 1 root root     0 Jun 14 09:50 web-1.txt

$ k exec -ti web-1 -- sh
# cd usr/share/nginx/html
# ls -la
total 24
drwxr-xr-x 3 root root  4096 Jun 14 09:50 .
drwxr-xr-x 4 root root  4096 Jul  1  2016 ..
drwx------ 2 root root 16384 Jun 14 09:47 lost+found
-rw-r--r-- 1 root root     0 Jun 14 09:50 web-1.txt
```

#### Environment:
- Kubernetes version (use `kubectl version`): v1.18.18

  The affected K8s versions are:
  - 1.18.16 - 1.18.19 (1.18 reached EoL)
  - 1.19.8 - 1.19.11 (to be fixed in 1.19.12 with backport of https://github.com/kubernetes/kubernetes/pull/101737)
  - 1.20.2 - 1.20.7 (to be fixed in 1.20.8 with backport of https://github.com/kubernetes/kubernetes/pull/101737)
  - 1.21.0 - 1.21.1 (to be fixed in 1.21.2 with backport of https://github.com/kubernetes/kubernetes/pull/101737)

- Cloud provider or hardware configuration: AWS

- OS (e.g: `cat /etc/os-release`):

```
PRETTY_NAME="Garden Linux 184.0"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://gardenlinux.io/"
SUPPORT_URL="https://github.com/gardenlinux/gardenlinux"
BUG_REPORT_URL="https://github.com/gardenlinux/gardenlinux/issues"
GARDENLINUX_FEATURES=base,aws,_nopkg,_prod,_readonly,_slim,base,cloud,gardener,server
GARDENLINUX_VERSION=184.0
GARDENLINUX_COMMIT_ID=e5a1576deded4cb778b178ed279c38138e384f63
VERSION_CODENAME=bullseye
```

- Kernel (e.g. `uname -a`):

```
Linux <omitted> 5.4.0-5-cloud-amd64 #1 SMP Debian 5.4.68-1 (2020-09-28) x86_64 GNU/Linux
```

- Install tools:

  Gardener: v1.23.2
  gardener-extension-provider-aws: v1.24.1
  aws-ebs-csi-driver - v0.10.1

- Network plugin and version (if this is a network-related bug): Not applicable
- Others:

**Comment by justaugustus** [comment]
cc: @kubernetes/release-engineering 
ref: https://kubernetes.slack.com/archives/CJH2GBF7Y/p1623702854323900?thread_ts=1623453145.304800&cid=CJH2GBF7Y

**Comment by wongma7** [comment]
> IMO this is a data race in K8s and this issue can arise with manual detach from the cloud provider (that's why it is not strictly related to issue 1.). We also saw that this issue can arise when only certain AWS machine types are used because we saw that the block device naming on for example m5.xlarge and x1.16xlarge is different.

can you elaborate on the block device naming issue? The manual detach case I am not too concerned about, kubelet has to trust that it has been provided a stable `devicePath:volume` mapping, any manual manipulation of that outside of kcm/csi attacher during the mount process is undefined behavior 

I am not fully understanding how the volume got detached from KCM, I will read the relevant PRs and code but appreciate any guidance. The volume is migrated, so why is KCM detaching it? Even with migration aside, attach detach controller should see that vol-1 is desired by pod-1, why did it detach it?

---



For my  reference:
in-tree cloudprovider device allocator: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/device_allocator.go
CSI device allocator/manager: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/25edc9779a6bf8608902908473d3868e05b6efb8/pkg/cloud/devicemanager/manager.go#L102

**Comment by wongma7** [comment]
> The manual detach case I am not too concerned about, kubelet has to trust that it has been provided a stable devicePath:volume mapping, any manual manipulation of that outside of kcm/csi attacher during the mount process is undefined behavior

We should use https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html#identify-nvme-ebs-device to verify the mapping before calling MountDevice

**Comment by ialidzhikov** [comment]
> can you elaborate on the block device naming issue?

On a `m5.xlarge` host, the output of `lsblk` is:

```
# lsblk
NAME        MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
nvme0n1     259:0    0  20G  0 disk
|-nvme0n1p1 259:5    0   1M  0 part
|-nvme0n1p2 259:6    0  16M  0 part /boot/efi
|-nvme0n1p3 259:7    0   1G  0 part /usr
`-nvme0n1p4 259:8    0  19G  0 part /
nvme1n1     259:1    0   1G  0 disk /var/lib/kubelet/pods/5754f696-05a1-46a5-857f-6d5c0bbcab4a/volumes/kubernetes.io~csi/pv-foo-f68918b6-bf55-451c-af47-84d05597c960/mount
nvme2n1     259:2    0   1G  0 disk /var/lib/kubelet/pods/0008954b-5286-448b-8889-e108b68b0649/volumes/kubernetes.io~csi/pv-foo-01f04ea4-b119-44c2-857c-2b0520057409/mount
```


When you try out the above steps to reproduce against a `m5.xlarge` host, at `step 12.` Pod `web-0` fails "steal" other Pod's PVC and fails to start with:

```
$ k describe po web-0

Events:
  Warning  FailedMount             25s (x8 over 89s)    kubelet                  MountVolume.MountDevice failed for volume "pv-foo-f68918b6-bf55-451c-af47-84d05597c960" : rpc error: code = Internal desc = Failed to find device path /dev/xvdba. nvme path "/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol1234" not found
```

On a `x1.16xlarge` host, the output of `lsblk` is:


```
# lsblk
NAME    MAJ:MIN   RM SIZE RO TYPE MOUNTPOINT
xvda    202:0      0  20G  0 disk
|-xvda1 202:1      0   1M  0 part
|-xvda2 202:2      0  16M  0 part /boot/efi
|-xvda3 202:3      0   1G  0 part /usr
`-xvda4 202:4      0  19G  0 part /
xvdba   202:13312  0   1G  0 disk /var/lib/kubelet/pods/a1ea9531-46ca-4859-9478-7a21e72069be/volumes/kubernetes.io~csi/pv-foo-f68918b6-bf55-451c-af47-84d05597c960/mount
xvdbb   202:13568  0   1G  0 disk /var/lib/kubelet/pods/c74cecc8-bdeb-4138-a5f9-0793f6703288/volumes/kubernetes.io~csi/pv-foo-01f04ea4-b119-44c2-857c-2b0520057409/mount
```

You can see that the device name is derived from the VolumeAttachment `.status.attachmentMetadata.devicePath`

When you try out the above steps to reproduce against a  `x1.16xlarge` host, you should end up with Pod `web-0` that "steals" other Pod's PVC and succeeds to start.

---

> The manual detach case I am not too concerned about, kubelet has to trust that it has been provided a stable devicePath:volume mapping, any manual manipulation of that outside of kcm/csi attacher during the mount process is undefined behavior

Manual detach is only used in the "artificial" steps to reproduce to demonstrate the race condition. In the real issue there was a detach from kube-controller-manager that is actually the trigger for the all mess. :)

---

> I am not fully understanding how the volume got detached from KCM, I will read the relevant PRs and code but appreciate any guidance. The volume is migrated, so why is KCM detaching it? Even with migration aside, attach detach controller should see that vol-1 is desired by pod-1, why did it detach it?

I think the description in https://github.com/kubernetes/kubernetes/pull/101423 is a good starting point. I think that this regression is introduced with backports of https://github.com/kubernetes/kubernetes/pull/98224. @msau42 and @Jiawei0227 can jump in and provide even more techinical details about this issue.

We definitely see this detach from kube-controller-manager logs. 12:31:29 - first, the volume is marked as "uncertain"; 12:32:02 DetachVolume starts; 12:34:01 DetachVolume succeeds.

kube-controller-manager logs:

```
2021-06-05 12:31:29 {"log":"Marking volume attachment as uncertain as volume:\"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\" (\"worker\") is not attached (Detached)","pid":"1","severity":"INFO","source":"attach_detach_controller.go:757"}
2021-06-05 12:32:02 {"log":"attacherDetacher.DetachVolume started for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\") on node \"worker\" ","pid":"1","severity":"INFO","source":"reconciler.go:219"}
2021-06-05 12:34:01 {"log":"DetachVolume.Detach succeeded for volume \"pvc-7e1ae9ef-2fb2-4e41-a1b4-a3a4d92edc59\" (UniqueName: \"kubernetes.io/aws-ebs/aws://eu-central-1a/vol-1\") on node \"worker\" ","pid":"1","severity":"INFO","source":"operation_generator.go:472"}
```

We confirmed this detach request even from AWS CloudTrail logs.

**Comment by wongma7** [comment]
Thanks, the nvme/nitro instance (like m5) not being affected by this issue makes sense because we already do use volume ID to decide what nvme volume to mount, as opposed to using whatever inaccurate devicePath/block device name got set in volumeattachment. So my previous comment that we should verify nvme mapping before mounting can be disregarded, we already do it for nitro instances. ref: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/25edc9779a6bf8608902908473d3868e05b6efb8/pkg/driver/node_linux.go#L53  (in-tree: https://github.com/kubernetes/kubernetes/blob/d01a5cae9cd819fa1038430f272eba8272b8ce6f/pkg/volume/awsebs/aws_util.go#L237)

So assuming that  https://github.com/kubernetes/kubernetes/pull/101737 fixes this specific KCM race, the remaining work is to see how we can mitigate potential future races?
1. for NVME volumes: decide if the volume ID pre-mount check (linked above) is sufficient? IMO yes. This happens in NodeStageVolume. The check is not perfect, there is always a slim possibility the volume gets detached between when the driver calls findDevicePath and when it calls FormatAndMount, but I think this is a possibility we have to live with, the driver can't control detaches happening outside its purview .
2. for non-NVME volumes: determine if a similar check is even possible. I asked AWS EBS, am waiting to hear back.

(I understand now why KCM is detaching and why https://github.com/kubernetes/kubernetes/pull/101737 fixes it... Attach detach controller translates before adding volume to desired state of world, but without https://github.com/kubernetes/kubernetes/pull/101737 the "uncertain volumes" routine didn't translate before marking volumes as uncertain in actual state of world, so controller detached them)
